VM Overview - set boot order

yaacov opened the pull request:
~Depends on - https://github.com/openshift/console/pull/3558~ - merged
~Depends on - https://github.com/openshift/console/pull/3560~ - merged

-----------

Add possibility to change boot order

  - [x] modal view.
  - [x] overview summary view .
  - [x] Implement for vms an vm templates.
  - [x] Implement keyboard control for a11y.
  - [x] implement update k8s on submit.
  - [x] handle negative flow.
  - [x] handle object changed in back-end.

Jira:
https://jira.coreos.com/browse/CNV-2907
https://jira.coreos.com/browse/CNV-2908
https://jira.coreos.com/browse/CNV-2909

Design:
https://github.com/openshift/openshift-origin-design/blob/master/web-console/knikubevirt/vm-details/vm-boot-order/vm-boot-order.md

Screenshots:

Change boot order:
![Peek 2019-12-11 10-57](https://user-images.githubusercontent.com/2181522/70606988-58000800-1c06-11ea-83f1-c86a93d0ae15.gif)

Empty state:
![Peek 2019-12-11 10-58](https://user-images.githubusercontent.com/2181522/70606990-58000800-1c06-11ea-9e37-c6b7cb6c1550.gif)

Update boot order from another process:
![Peek 2019-12-11 11-05](https://user-images.githubusercontent.com/2181522/70606991-58000800-1c06-11ea-9029-4f083e4fe0c7.gif)

yaacov commented:
@mareklibra @suomiy @rawagner @glekner @irosenzw please review

yaacov commented:
cc:// @Ranelim

yaacov commented:
cc:// @amirfefer

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/vm-boot-order-modal/components/dnd-list/dnd-list.tsx:
```diff
@@ -0,0 +1,61 @@
+}) => (
+  <DndProvider backend={HTML5Backend}>
+    <DataList aria-label="sorted list">
+      {[...Array(size).keys()].map((index) => (
```
you can use `_.range` instead

yaacov replied:
:tada:  thanks :+1:, looked for something like that, and missed it  :-)

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/vm-boot-order-modal/components/dnd-list/dnd-list.tsx:
```diff
@@ -0,0 +1,61 @@
+  <DndProvider backend={HTML5Backend}>
+    <DataList aria-label="sorted list">
+      {[...Array(size).keys()].map((index) => (
+        <DNDRow
```
rows should have keys

yaacov replied:
added :+1:

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/vm-boot-order-modal/components/dnd-list/dnd-list.tsx:
```diff
@@ -0,0 +1,61 @@
+        />
+      ))}
+      {canAddItem && (
+        <DNDRowAddItem
```
add row also

yaacov replied:
added :+1:

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/vm-boot-order-modal/components/dnd-list/dnd-list.tsx:
```diff
@@ -0,0 +1,61 @@
+  addItemSetEditMode,
+}) => (
+  <DndProvider backend={HTML5Backend}>
+    <DataList aria-label="sorted list">
```
this does not tell much and should be probably passed from above

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/vm-boot-order-modal/components/dnd-list/dnd-row.tsx:
```diff
@@ -0,0 +1,81 @@
+const DNDRowDragHandleStyle = { cursor: 'move' };
+const DNDRowDeleteStyle = { cursor: 'pointer' };
+
+export const DNDRow: React.FC<DNDRowProps> = ({ Component, icon, index, onDelete, onMove }) => {
```
these components should be reusable and not live under `vm-boot-order-modal`

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/vm-boot-order-modal/components/dnd-list/dnd-row.tsx:
```diff
@@ -0,0 +1,81 @@
+    <div ref={(node) => preview(drop(node))} style={{ opacity }}>
+      <DataListItem
+        aria-labelledby="sorted list item"
+        className="kubevirt-boot-order-modal__list-item"
```
I can't find all of these css classes anywhere

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/vm-boot-order-modal/components/dnd-list/dnd-row.tsx:
```diff
@@ -0,0 +1,81 @@
+  return (
+    <div ref={(node) => preview(drop(node))} style={{ opacity }}>
+      <DataListItem
+        aria-labelledby="sorted list item"
```
sorted?

yaacov replied:
changed to dnd

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/vm-boot-order-modal/boot-order.tsx:
```diff
@@ -0,0 +1,175 @@
+  const addItemRow: React.FC<any> = AddItemRowFactory(state.options, addByKey);
+
+  return (
+    <Modal
```
Could you please extract the modal part to a separate component (due to the reusabilty for the VM wizard)?

Something like a `disk-modal.tsx` and `disk-modal-enhanced.tsx`. The difference is that boot order will not be represented as a modal when inside the VM wizard. It will be a second advanced tab (likely a followup PR to this one).

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/vm-boot-order-modal/boot-order.tsx:
```diff
@@ -0,0 +1,175 @@
+    setState({ sources, options: state.options, isEditMode: false, changed: true });
+  };
+
+  const submit = async (e) => {
```
please can you also extract the request? Will not be used in the VM wizard

suomiy replied:
just a generic callback

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/vm-boot-order-modal/components/dnd-list/dnd-list.tsx:
```diff
@@ -0,0 +1,61 @@
+
+import './dnd-list.scss';
+
+export const DNDList: React.FC<DNDListProps> = ({
```
IMO, I don't think this component is really necessary. We could just use the body inplace, because the number of props is not that much different from the number of lines of the body

yaacov replied:
Shouldn't this be reusable for other components ?

suomiy replied:
I think one can put this together in the place of need very easily. It is just that this component is heavily customized (props) and one could just use the plain components instead

yaacov commented on frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx:
```diff
@@ -9,6 +9,7 @@ import { VMTemplateLink } from '../vm-templates/vm-template-link';
 import { getBasicID, prefixedID } from '../../utils';
 import { vmDescriptionModal, vmFlavorModal } from '../modals';
 import { VMCDRomModal } from '../modals/cdrom-vm-modal';
+import { VMBootOrderModal } from '../modals/vm-boot-order-modal';
```
yes :+1: , before that it will need some twiking with the k8s save function, so I can't use it their yet.

will add it to the items to do in the description above.

suomiy replied:
have you tried `getVMLikePatches()` ? or is there some other problem?

yaacov replied:
hmmm :-) didn't know about that ... looking ...

yaacov replied:
using the `getVMLikePatches` method :+1:

should work for templates now too, didn't check  :-)

yaacov replied:
done, works for templates as expected :-)

yaacov commented on frontend/packages/kubevirt-plugin/src/components/modals/vm-boot-order-modal/boot-order.tsx:
```diff
@@ -0,0 +1,175 @@
+
+  const onDelete = (index: number) => {
+    const item = state.sources[index];
+    delete item.value.bootOrder;
```
using: https://github.com/openshift/console/pull/3436#discussion_r347310059

yaacov commented on frontend/packages/kubevirt-plugin/src/components/modals/vm-boot-order-modal/boot-order.tsx:
```diff
@@ -0,0 +1,175 @@
+    const item = state.sources[index];
+    return <p>{`${item.value.name} (${item.typeLabel})`}</p>;
+  };
+  const addItemRow: React.FC<any> = AddItemRowFactory(state.options, addByKey);
```
alphabetically ?

suomiy replied:
could be, + maybe also by device first?

yaacov replied:
:+1:

yaacov replied:
sorted :cake:

yaacov commented on frontend/packages/kubevirt-plugin/src/components/modals/vm-boot-order-modal/add-item-row-factory.tsx:
```diff
@@ -0,0 +1,23 @@
+  <FormSelect
+    value=""
+    id="boot-order-select"
+    className="kubevirt-boot-order-modal__input pf-m-9-col-on-md"
```
cleaned up the unused class names :+1:

yaacov commented on frontend/packages/kubevirt-plugin/src/components/modals/vm-boot-order-modal/add-item-row-factory.tsx:
```diff
@@ -0,0 +1,23 @@
+import * as React from 'react';
+import { FormSelect, FormSelectOption } from '@patternfly/react-core';
+
+export const AddItemRowFactory = (
```
moved inside :-)

yaacov commented on frontend/packages/kubevirt-plugin/src/components/modals/vm-boot-order-modal/boot-order.tsx:
```diff
@@ -0,0 +1,175 @@
+export const BootOrderModal = withHandlePromise((props: BootOrderModalProps) => {
+  const { vmLikeEntity, handlePromise, onClose } = props;
+  const [state, setState] = React.useState<BootOrderModalState>({
+    sources: getBootableDevices(vmLikeEntity),
```
done :exclamation:

yaacov commented:
@suomiy  
a - moved components to components dir
b - separated the component from the modal
c - updated the template view.

still a WIP because it fails linting ?! can't find out why :-(
```
/home/yzamir/go/src/github.com/openshift/console/frontend/public/components/utils/file-input.tsx
  4:22  error  ConnectDropTarget not found in 'react-dnd'  import/named
  4:41  error  DropTargetMonitor not found in 'react-dnd'  import/named
```

yaacov commented:
@mareklibra @suomiy removed the wip, please re-review.

yaacov commented:
Update:
- tests work again, needed jest `moduleNameMapper` settings for react-dnd
- lint works again, thanks to @suomiy :+1: 

@suomiy raised the question of regressions due to bumping react-dnd version:
I tested the the components that use react-dnd [ yaml-editor and store-class-name-input ] , they seam to work the same way they did with older version of react-dnd.

yaacov commented:
/test e2e-gcp-console

suomiy commented on frontend/packages/kubevirt-plugin/src/components/boot-order/boot-order.tsx:
```diff
@@ -0,0 +1,144 @@
+    const sources = [...devices.sources, item];
+    const options = _.filter(
+      devices.options,
+      (option) => `${option.type}-${option.value.name}` !== key,
```
could we have some util function for this string generation?

yaacov replied:
done

suomiy commented on frontend/packages/kubevirt-plugin/src/components/boot-order/boot-order.tsx:
```diff
@@ -0,0 +1,144 @@
+
+    setDevices({
+      sources,
+      options: _.orderBy(options, ['type', 'value.name']),
```
orderBy could be done only just when rendering

yaacov replied:
done !

yaacov commented:
/test e2e-gcp-console

yaacov commented on frontend/packages/kubevirt-plugin/src/components/boot-order/add-item-row-factory.tsx:
```diff
@@ -0,0 +1,31 @@
+  >
+    <FormSelectOption label={message} value="" />
+    {options.map((option) => (
+      <FormSelectOption
```
done !

yaacov replied:
done

yaacov commented on frontend/packages/kubevirt-plugin/src/components/boot-order/boot-order.tsx:
```diff
@@ -0,0 +1,144 @@
+//   ...
+//   <BootOrder devices={devices} setDevices={setDevices} />
+export const BootOrder = ({ devices, setDevices }: BootOrderProps) => {
+  const [state, setState] = React.useState<BootOrderState>({
```
done

yaacov commented on frontend/packages/kubevirt-plugin/src/components/boot-order/boot-order.tsx:
```diff
@@ -0,0 +1,144 @@
+// 2 - regular mode        - show a list of sources, with a button to enter edit mode.
+// 3 - edit mode           - show a list of sources, with a select line for new source.
+//
+// Props <BootOrderProps>:
```
ahh , not sure about that :-)
removing docs, until interface is stable

yaacov commented on frontend/public/components/droppable-edit-yaml.tsx:
```diff
@@ -18,7 +18,7 @@ const boxTarget = {
 };
 
-const EditYAMLComponent = DropTarget(NativeTypes.FILE, boxTarget, (connectObj, monitor) => ({
+const EditYAMLComponent: any = DropTarget(NativeTypes.FILE, boxTarget, (connectObj, monitor) => ({
```
yes,

in v2.x `DropTarget` returned type `any` and now it returns a type that is more precise but is missing the custom fields of `EditYAML` defined in:

https://github.com/openshift/console/blob/master/frontend/public/components/edit-yaml.jsx#L77`

If we do not override the type returned we will get:

```
RROR in /home/yzamir/go/src/github.com/openshift/console/frontend/public/components/droppable-edit-yaml.tsx
(85,10): Type '{ obj: string; fileUpload: string | ArrayBuffer; error: string; onDrop: (item: any, monitor: any) => void; children?: ReactNode; }' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<Pick<Pick<any, never>, never>, any, any>> & Readonly<Pick<Pick<any, never>, never>> & Readonly<{ children?: ReactNode; }>'.
  Property 'obj' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<Pick<Pick<any, never>, never>, any, any>> & Readonly<Pick<Pick<any, never>, never>> & Readonly<{ children?: ReactNode; }>'.

```

yaacov commented on frontend/packages/kubevirt-plugin/src/components/modals/boot-order-modal/boot-order-modal.tsx:
```diff
@@ -0,0 +1,90 @@
+      event.preventDefault();
+
+      // Reset devices bootOrder using the list index.
+      const sources = _.cloneDeep(devices.sources).map((source, index) => {
```
done

yaacov commented on frontend/packages/kubevirt-plugin/src/selectors/vm/combined.ts:
```diff
@@ -37,3 +38,31 @@ export const findConversionPod = (vm: VMKind, pods: PodKind[]) => {
+export const getDevices = (vm: VMLikeEntityKind) => {
+  const disks = getDisks(asVM(vm)).map((disk) => ({
+    type: 'disk',
+    typeLabel: _.has(disk, 'cdrom') ? 'CD-ROM' : 'Disc',
```
done

yaacov commented:
update, now using children.

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

suomiy commented on frontend/packages/kubevirt-plugin/src/components/boot-order/boot-order.tsx:
```diff
@@ -0,0 +1,115 @@
+} from './constants';
+
+export const BootOrder = ({ devices, setDevices }: BootOrderProps) => {
+  const { sources, options } = {
```
is there a need for wrapping these inside an extra object?

suomiy commented on frontend/packages/kubevirt-plugin/src/constants/vm/storage/disk-type.ts:
```diff
@@ -1,6 +1,12 @@
+import * as _ from 'lodash';
 import { ValueEnum } from '../../value-enum';
 
+const typeLabels = {
```
the label could be also put into the constructor

suomiy commented on frontend/packages/kubevirt-plugin/src/components/boot-order/boot-order.tsx:
```diff
@@ -0,0 +1,115 @@
+    const newSources = [...sources, item];
+    const newOptions = _.filter(options, (option) => deviceKey(option) !== key);
+
+    setDevices([...newSources, ...newOptions]);
```
can we just map devices and replace the device we need instead?

suomiy commented on frontend/packages/kubevirt-plugin/src/components/boot-order/boot-order.tsx:
```diff
@@ -0,0 +1,115 @@
+  const addByKey = (key: string): void => {
+    const item = _.find(options, (option) => deviceKey(option) === key);
+    // Add boot order to device.
+    item.value.bootOrder =
```
please do not modify the original device, but make at least a shallow copy

suomiy commented on frontend/packages/kubevirt-plugin/src/components/boot-order/boot-order.tsx:
```diff
@@ -0,0 +1,115 @@
+    // Remove boot order from device.
+    delete item.value.bootOrder;
+
+    const newSources = [...sources.slice(0, index), ...sources.slice(index + 1)];
```
is this necessary? shouldn't it be enough to just remove the bootOrder?

suomiy commented on frontend/packages/kubevirt-plugin/src/components/boot-order/boot-order.tsx:
```diff
@@ -0,0 +1,115 @@
+    const item = sources[index];
+
+    // Remove boot order from device.
+    delete item.value.bootOrder;
```
\+ another mutation

suomiy commented on frontend/packages/kubevirt-plugin/src/components/dnd-list/dnd-data-list-item.tsx:
```diff
@@ -0,0 +1,83 @@
+
+  return (
+    <div ref={(node) => preview(drop(node))} style={{ opacity }}>
+      <DataListItem aria-labelledby="dnd list item">
```
`labelledby` what element?

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/boot-order-modal/boot-order-modal.tsx:
```diff
@@ -0,0 +1,70 @@
+
+    const submit = async (event) => {
+      event.preventDefault();
+
```
maybe we should "relax" the bootorder before the request? It could be higher depending on remove/add of devices

yaacov replied:
I added a relax code to `delete`, `add` and `move` already leave the device boot order relaxed.
so now all operations leave the list relaxed.

suomiy commented on frontend/packages/kubevirt-plugin/src/components/boot-order/boot-order.tsx:
```diff
@@ -0,0 +1,115 @@
+        />
+      ) : (
+        <DNDDataList ariaLabel="vm boot order">
+          {_.range(sources.length).map((index) => (
```
you can use simple map here and get the index as a second parameter

yaacov commented:
/test e2e-gcp-console

yaacov commented:
@suomiy update:
a - fixed the boot order callbacks.
b - added a boot-order-summary component using the design.

@mareklibra 
a - I moved some selectors from using the `web-ui-components` version to use the internal `selectors` version, PTAL. 
 
yet-to-do:
a- the firehose thing

yaacov commented:
/test images
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

suomiy commented on frontend/packages/kubevirt-plugin/src/components/boot-order/boot-order.tsx:
```diff
@@ -0,0 +1,126 @@
+
+    const bootOrder = sources.length !== 0 ? sources[sources.length - 1].value.bootOrder + 1 : 1;
+    newDevices.find((device) => deviceKey(device) === key).value.bootOrder = bootOrder;
+
```
\+ relax here - the array might not have been relaxed to begin with

yaacov replied:
done

suomiy commented on frontend/packages/kubevirt-plugin/src/components/boot-order/boot-order.tsx:
```diff
@@ -0,0 +1,126 @@
+    delete newDevices.find((device) => deviceKey(device) === key).value.bootOrder;
+
+    // Reset newDevices bootOrder using the list index.
+    _.filter(newDevices, (device) => device.value.bootOrder).forEach((source, i) => {
```
we might not need the _.cloneDeep if we use `devices.map` and `_.merge` instead

yaacov replied:
I think `_,marge` creates a new object like `_.cloneDeep` does ?

suomiy replied:
`_.merge({}, originalDevice, { value: { bootOrder: 3}})` will just create a shallow copy of the outer wrapper object  and shallow copy of value, but it should keep the other references intact. Thus it should be better optimized than cloneDeep which creates everything from scratch.

suomiy commented on frontend/packages/kubevirt-plugin/src/components/boot-order/boot-order.tsx:
```diff
@@ -0,0 +1,126 @@
+              onMove={onMove}
+              key={`${dataListID}-${deviceKey(source)}`}
+            >
+              <Text component={TextVariants.p}>{deviceLabel(source)}</Text>
```
shouldn't the `DNDDataListItem` be aria-labelledby this `Text` component rather than the whole list?

yaacov replied:
make sense to me too :-)

suomiy commented on frontend/packages/kubevirt-plugin/src/components/boot-order/boot-order.tsx:
```diff
@@ -0,0 +1,126 @@
+            </DNDDataListItem>
+          ))}
+          <DNDDataListItem
+            dataListID={dataListID}
```
same here

yaacov replied:
done

suomiy commented on frontend/packages/kubevirt-plugin/src/constants/vm/storage/disk-type.ts:
```diff
@@ -1,11 +1,16 @@
   static readonly CDROM = new DiskType('cdrom');
   static readonly FLOPPY = new DiskType('floppy');
   static readonly LUN = new DiskType('lun');
+  static readonly typeLabels = {
```
```suggestion
  private readonly label: string;
```


and then please add it to the constructor

suomiy commented on frontend/packages/kubevirt-plugin/src/components/boot-order/add-item-form-select.tsx:
```diff
@@ -0,0 +1,33 @@
+}) => (
+  <FormSelect
+    value=""
+    id="boot-order-select"
```
maybe we should pass these as a parameter or a default parameter?

yaacov replied:
now a required parameter :+1:

yaacov commented:
/test e2e-gcp-console

matthewcarleton commented:
continuing conversation around naming from [here](https://github.com/openshift/console/pull/3560#discussion_r351312881)
@suomiy  The naming comes from the CD-ROMs section where we distinguish the drives by Drive 1, Drive 2. Whatever we do here should align with how CD-ROMs are handled. I am ok with updating this if we think it's clear to the user.

yaacov commented:
> To avoid confusion with the format you suggested:

> 1. Drive-2
> 2. Drive-1

> We could drop the Drive-1

> 1. container-image.iso (CD-ROM) 

> prefix or append the drive number to the (CD-ROM) so it could look like :

> 1. container-image.iso (CD-ROM drive 1)

> Another option

> 1. container-image.iso (CD-ROM 1)

My vote is for dropping the drive number. My reasoning is that assigning a drive number is misleading, the VM does not have CD drives ( or drive numbers ) the numbers we show as drives are arbitrary numbers, it's the order we find the CD's on the yaml device list.

> I  suppose the only question left is do we align CD-ROMs to this format as well? 

My vote is to change the CD list too, it make sense to me to use the same label generating method, it will allow changing the format in one place if we choose to change it. [ cc @glekner ] 

> Maybe we can just leave the Drive-1 in the modal and not have it in the list of CD-ROMs.
I'm a little concerned with this approach it feels like the more info we provide the better the accessibility will be.

yaacov commented:
/test e2e-gcp-console

matthewcarleton commented:
@yaacov I'm ok with that. 
@glekner keep me honest, but I believe we added the Drive number because a CD can be ejected from the CD-ROM but the CD-ROM is still there. In that case you could have 
```
Drive 1 : container-image.iso
Drive 2: another-image.iso
```
Eject drive 1 but not remove the drive itself which would give you in the details
```
Drive 2: another-image.iso
```
When editing the drives you would still have Drive 1 available with the ability to attach another CD.
It's because of this that we added the drive count labels.  If we remove them we will then just rely on the contents of the drive to distinguish them. We'd still want to keep Drive 1, Drive 2 in the modal though because we need some way to label them.

yaacov commented:
/test e2e-gcp-console

suomiy commented:
seeing there are many problems with this, can we trace back a little bit?

why don't we just name the drives accordingly like drive-1 and drive-2 and show them in the order they appear in the YAML?

- we generate these names and they are not arbitrary
- we don't have to allow editing for them
- we can even do simple beautifying like drive-1 -> drive 1, even though I prefer just drive-1
- user can differentiate between their own drives from their own templates
- easier and consistent usage in boot order + other devices in other places

yaacov commented:
> why don't we just name the drives accordingly like drive-1 and drive-2 and show them in the order they appear in the YAML?

@suomiy I think I'm missing something, I do not know how to map "drives" to an entity I can see in the YAML , for example:
``` yaml
devices:
  rng: {}
  interfaces:
    - name: nic0
      model: virtio
      bootOrder: 2
      masquerade: {}
  disks:
    - name: rootdisk
      bootOrder: 1
      disk:
        bus: virtio
    - name: mycd
      cdrom:
        bus: virtio
    - name: cloudinitdisk
      bootOrder: 3
      disk:
        bus: vi
```
I have `devices`, `disks` and `interfaces` how do we represent `drives` , do they refer to `volumes`?
``` yaml
volumes:
  - name: rootdisk
    containerDisk:
      image: fed
```

suomiy commented:
> @suomiy I think I'm missing something, I do not know how to map "drives" to an entity I can see in the YAML , for example:
> 
> I have `devices`, `disks` and `interfaces` how do we represent `drives` , do they refer to `volumes`?

not sure what you mean by that?

The disks are associated with volumes and the name is the key

 ```yaml
devices:
  disks:
    - name: my-cd-drive-1
      cdrom:
       bus: virtio
```

and

```yaml
volumes:
  - name: my-cd-drive-1
    containerDisk:
      image: fed
```

matthewcarleton commented:
@yaacov I'm testing this with a keyboard and I can't currently tab to each item. Is that something we can incorporate here? Here's a really good article on how to possibly implement this.
https://medium.com/salesforce-ux/4-major-patterns-for-accessible-drag-and-drop-1d43f64ebf09
please let us know how we can help further with this effort and what your thoughts are :)

yaacov commented:
updates:
a - added keyboard `arrowUp` / `arrowDown` / `minus` for item move and delete operations from keyboard.
b. moved the `add item` componenet out of the items list.

Screenshot of keyboard usage:
![Peek 2019-12-01 17-23](https://user-images.githubusercontent.com/2181522/69916146-ea164c80-145f-11ea-93cd-98c67625d56f.gif)

yaacov commented:
> I'm testing this with a keyboard and I can't currently tab to each item. Is that something we can incorporate here?

@matthewcarleton thanks for the link :+1:, keyboard is now active, but I left out visual cues (e.g. `:fucos` css) and labels for your design, ( tried some things but didn't like them :-) ).

> please let us know how we can help further with this effort

yes, help with labeling e.g. what element should have labels and what should be the text and css for focus element, will be very helpful.

> Does the VM need to be shut down to change the boot order?

No, we can change the boot order when ever we want.

> Should we alert the user that the VM needs to be shut down before the change takes effect?

Boot order change will take affect on next boot, so IMHO it makes sense to alert user that changes will take effect on next reboot.

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
@suomiy hi, modal is now connected to changes in redux, it uses the changes to the vm prop to re evaluate the devices list:
![Peek 2019-12-02 16-54](https://user-images.githubusercontent.com/2181522/69969480-2a8bce00-1525-11ea-9b51-f516ac28fff6.gif)

yaacov commented:
/test images
/test e2e-gcp-console

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/boot-order-modal/boot-order-modal.tsx:
```diff
@@ -0,0 +1,93 @@
+  React.useEffect(() => {
+    const updatedDevices = getDevices(vmLikeEntity);
+
+    setDevices(updatedDevices);
```
please look again at https://github.com/openshift/console/pull/3436#discussion_r348083655

we should ignore the changes of the boot order from somebody else. The newest bootorder should come from the dialog. I think user is not interested in boot order changes if s/he wishes it to change.

yaacov replied:
ok ... will change to logic here ...

yaacov replied:
update:
a - inform users if boot order changed outside.
b - on submit, only change the boot order, and only for currently available devices.

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/boot-order-modal/boot-order-modal.tsx:
```diff
@@ -0,0 +1,93 @@
+  ];
+
+  return (
+    <Modal
```
we are missing promise progress and error here Shouldn't we align this with console dialogs?

yaacov replied:
> Shouldn't we align this with console dialogs?

console dialogs also ignore promises ( or rather don't do anything with them ) and leave the catching to the user.

for example in [disk-modal-enhanced.tsx](https://github.com/openshift/console/blob/master/frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal-enhanced.tsx#L41) we use console dialog with the same logic as here:
wait for promise to fulfil => then close dialog => ignore errors. 

> missing promise progress

we wait for promise to fulfil before closing the dialog ... since we do not handle negative flows it's the best we can do, until we start catching errors/timeouts.

> and error here

yes, we should handle errors, this will also fix the promise progress thing :-)

suomiy replied:
> > Shouldn't we align this with console dialogs?
> 
> console dialogs also ignore promises ( or rather don't do anything with them ) and leave the catching to the user.

I think this feature is in a lot of dialogs and is quite useful

> 
> for example in [disk-modal-enhanced.tsx](https://github.com/openshift/console/blob/master/frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal-enhanced.tsx#L41) we use console dialog with the same logic as here:
> wait for promise to fulfil => then close dialog => ignore errors.
> 


that is not true, disk-modal-enhanced covers all cases. Please see disk-modal and `withHandlePromise` wrapper with `handlePromise`. Can we use something like that?


> > missing promise progress
> 
> we wait for promise to fulfil before closing the dialog ... since we do not handle negative flows it's the best we can do, until we start catching errors/timeouts.
> 
> > and error here
> 
> yes, we should handle errors, this will also fix the promise progress thing :-)

+1 we should handle all the cases

yaacov replied:
update:
a - wait for patch to finish or fail, on ok -> close modal.
b - on error, inform user and leave modal open.
c - on outside change, inform users, let user manually reload devices.

suomiy replied:
can we rather use `withHandlePromise` like in all of our modals and a lot of console ones? 

- we get a design consistency
- error and progress handling for free

suomiy commented on frontend/packages/kubevirt-plugin/src/components/dnd-list/dnd-data-list-item.tsx:
```diff
@@ -0,0 +1,105 @@
+  // ArrowUp:   move item one order index down.
+  // ArrowDown: move item one order index up.
+  // '-':       delete an item.
+  const onKeyDown = (e: React.KeyboardEvent) => {
```
I am sorry, but it did not work for me. I think basic dnd is also broken

suomiy commented on frontend/packages/kubevirt-plugin/src/components/boot-order/add-device.tsx:
```diff
@@ -0,0 +1,51 @@
+  // Add a bootOrder to a device by key, item key is "<type>->name>".
+  const addByKey = (key: string): void => {
+    const newDevices = _.cloneDeep(devices);
+    // Make bootOrder none null, any number will be ok.
```
how does this work?

suomiy commented:
> @suomiy hi, modal is now connected to changes in redux, it uses the changes to the vm prop to re evaluate the devices list:

there are few problems with this approach and current implementation - please see the review

yaacov commented on frontend/packages/kubevirt-plugin/src/components/boot-order/add-device-form-select.tsx:
```diff
@@ -0,0 +1,49 @@
+  options: BootableDeviceType[];
+  label: string;
+  onDelete: () => void;
+  addByKey: (key: string) => void;
```
make sense to me too :+1: , done

yaacov commented on frontend/packages/kubevirt-plugin/src/components/boot-order/add-device-form-select.tsx:
```diff
@@ -0,0 +1,49 @@
+  </>
+);
+
+export type AddDeviceFormSelectType = {
```
fixed :heart:

yaacov commented on frontend/packages/kubevirt-plugin/src/components/boot-order/add-device-form-select.tsx:
```diff
@@ -0,0 +1,49 @@
+    <FormSelect
+      value=""
+      id={id}
+      // addByKey moves items from the options list to the sources list,
```
nice, moved :cloud:

yaacov commented on frontend/packages/kubevirt-plugin/src/components/boot-order/empty.tsx:
```diff
@@ -0,0 +1,42 @@
+} from '@patternfly/react-core';
+
+// Display and empty with a Call to add new source if no sources are defined.
+export const Empty: React.FC<EmptyProps> = ({
```
I was wondering about that ... sure, naming is hard, ( also renamed the none-summary empty to match ) :smile:

yaacov commented:
update, added alert:

@suomiy @matthewcarleton please review.

Updated outside:
![Peek 2019-12-03 11-33](https://user-images.githubusercontent.com/2181522/70038610-37f69600-15c1-11ea-99fb-a0e8bef2a914.gif)

Error saving:
![Peek 2019-12-03 11-35](https://user-images.githubusercontent.com/2181522/70038625-3cbb4a00-15c1-11ea-9f2a-58e0a0e2b577.gif)

yaacov commented:
/test e2e-gcp-console
/test images

suomiy commented:
> update, added alert:
> 

maybe we could add a refresh button, upon which would the boot order reset to the recent change? - something like `refresh` with spinner2 from https://www.patternfly.org/v4/design-guidelines/styles/icons

@yaacov @matthewcarleton wdyt?


I would prefer the original error to propagate

yaacov commented:
> @yaacov @matthewcarleton wdyt?

I like this :+1:

> I would prefer the original error to propagate

will look how to do it ...

yaacov commented:
Removed WIP:
a - show alerts on negative flow.
b - on submit update only bootOrder.
c - show info if opbject is changed outside.
d - has kbd input, except for select dropdown :-(

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

matthewcarleton commented:
@yaacov for the alerts is it possible to provide more info on why it was failing?
@suomiy are you saying add a try again action in the fail alert?

yaacov commented:
> @yaacov for the alerts is it possible to provide more info on why it was failing?

I'm sending all the info I have, in this screenshot the error is "Type Error: Network error ..":
 
![OKD](https://user-images.githubusercontent.com/2181522/70062981-c4b84880-15ef-11ea-8e5c-ad7482125a06.png)

> @suomiy are you saying add a try again action in the fail alert?

I think @suomiy refers to the `Reload` button, like the one we have in the edit YAML page:

![OKD(1)](https://user-images.githubusercontent.com/2181522/70063309-37292880-15f0-11ea-82fb-d42c7c7e46c7.png)

matthewcarleton commented:
ah thanks @yaacov I get it now :)
Would it make sense to add the reload action in the alert itself? Not 100% familiar with the YAML pattern but it seems a bit more a11y friendly and doesn't require us to update the general actions for the modal.

yaacov commented:
@matthewcarleton `reload` action now connected to the `updated` alert.
I also updated the screenshots in the PR description.

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test analyze

yaacov commented:
/test e2e-gcp-console

yaacov commented:
@suomiy I'm adding individual commits because I think it helps reviewing, do you think it's better to squash ?

yaacov commented:
/test analyze

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
squashed

yaacov commented:
/test e2e-gcp-console

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/boot-order-modal/boot-order-modal.tsx:
```diff
@@ -0,0 +1,156 @@
+  // Inform user on vmLikeEntity.
+  React.useEffect(() => {
+    // Check if bootOrder of current device changed.
+    if (!_.isEqualWith(initialDeviceList, getDevices(vmLikeEntity), _.isEqual)) {
```
what is the benefit of using isEqualWith here?

yaacov replied:
changed the logic here to only trigger on boot order changes

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/boot-order-modal/boot-order-modal.tsx:
```diff
@@ -0,0 +1,156 @@
+    if (isOpen) {
+      onReload();
+    }
+  }, [isOpen]); // eslint-disable-line react-hooks/exhaustive-deps
```
should probably reference `onReload` as well

yaacov replied:
`onReload` changes when we get a new `vmLikeEntitiy`, this should only triger on `isOpen` changing to `true` ...

suomiy replied:
so the reference to `onReload` would be stale when it changes to true, no?

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/boot-order-modal/boot-order-modal.tsx:
```diff
@@ -0,0 +1,156 @@
+
+      <div className="kubevirt-boot-order-modal__alerts">
+        {showUpdatedAlert && (
+          <Alert variant="warning" isInline title="Boot order has been updated outside this flow.">
```
This isn't always true, as we detect all changes. E.g. just bus could have changed, while boot order stays untouched.

yaacov replied:
now trigger only on boot order changes

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/boot-order-modal/boot-order-modal.tsx:
```diff
@@ -0,0 +1,156 @@
+
+      <div className="kubevirt-boot-order-modal__alerts">
+        {showUpdatedAlert && (
+          <Alert variant="warning" isInline title="Boot order has been updated outside this flow.">
```
also, you can import `AlertVariant` from pfreact. And the error one should be error

yaacov replied:
done.

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/boot-order-modal/boot-order-modal.tsx:
```diff
@@ -0,0 +1,156 @@
+            Saving these changes will override any previously saved updates.
+            <br />
+            To see the updated order{' '}
+            <Button variant="link" isInline onClick={onReload}>
```
same for button variants

yaacov replied:
done.

rawagner commented:
/approve

yaacov commented on frontend/packages/kubevirt-plugin/src/components/boot-order/boot-order.tsx:
```diff
@@ -0,0 +1,115 @@
+
+  // Relax bootOrder and use setDevice to update the parent componenet.
+  const updateDevices = (newDevices: BootableDeviceType[]): void => {
+    _.filter(newDevices, (device) => device.value.bootOrder).forEach((source, i) => {
```
done.

yaacov commented on frontend/packages/kubevirt-plugin/src/components/modals/boot-order-modal/boot-order-modal.tsx:
```diff
@@ -0,0 +1,155 @@
+    event.preventDefault();
+
+    // Copy only bootOrder from devices to current device list.
+    const currentDevices = _.cloneDeepWith(getDevices(vmLikeEntity));
```
done.

yaacov commented:
/test e2e-gcp-console
/test images

yaacov commented:
/test e2e-gcp-console

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/boot-order-modal/boot-order-modal.tsx:
```diff
@@ -0,0 +1,183 @@
+    >
+      <BootOrder devices={devices} setDevices={setDevices} />
+
+      <div className="kubevirt-boot-order-modal__alerts">
```
can we reuse `ModalFooter` from `modals` ?

suomiy replied:
you can enhance it to fit your needs better if you need. Maybe also adding children to it if there will be big feature spread.

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/boot-order-modal/boot-order-modal.tsx:
```diff
@@ -0,0 +1,183 @@
+            {errorMessage}
+          </Alert>
+        )}
+        {inProgress && <Alert variant={AlertVariant.info} isInline title="Request in progress." />}
```
not sure if the user needs to know that :) loading icon should be enough

matthewcarleton commented on frontend/packages/kubevirt-plugin/src/components/modals/boot-order-modal/boot-order-modal.tsx:
```diff
@@ -0,0 +1,183 @@
+            isInline
+            title="Boot order has been updated outside this flow."
+          >
+            Saving these changes will override any previously saved updates.
```
ya good point. I'd say 
`Saving these changes will override any boot order previously saved.`

yaacov commented:
/test e2e-gcp-console
/test images

yaacov commented:
/test analyze

yaacov commented:
/test analyze

matthewcarleton commented:
@yaacov is the screen shots at the top of this the latest version?
If so, can we replace the primary action in the empty state with a secondary action?
Also, I'm seeing a blue "-" icon which should be `--pf-global--Color--100` when active/focus

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console
/test images

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
@matthewcarleton hi,

> @yaacov is the screen shots at the top of this the latest version?

updated all screenshots now :postal_horn: 

> can we replace the primary action in the empty state with a secondary action?

sure, done :unicorn: 

> Also, I'm seeing a blue "-" icon which should be --pf-global--Color--100 when active/focus

sure, done :parasol_on_ground:

yaacov commented:
/test e2e-gcp-console
/test analyze

yaacov commented:
/test analyze

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

suomiy commented:
/lgtm

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console
